Recursive Derives and Attributes

tadeohepperle opened the pull request:
When adding derives and attributes to a specific type, there is now an option to have them applied recursively to all dependencies of this type.

From the user side, the only thing that changed is that the `extend_for_type()` function on a `DerivesRegistry` that is part of the `TypegenSettings` now has an additional parameter `recursive: bool`.
Internally the `DerivesRegistry` maintains two: HashMaps mapping type paths to derives/attributes:
- `specific_type_derives: HashMap<syn::TypePath, Derives>`
- `recursive_type_derives: HashMap<syn::TypePath, Derives>`

When it comes to creating code and resolving recursive dependencies, these two maps are flattened out into a single `specific_type_derives: HashMap<syn::TypePath, Derives>`. This is done once, in the beginning of the code generation process, transforming the `DerivesRegistry` into a `FlatDerivesRegistry`. Only at this point knowledge of the type dependencies need to be known in form of a `PortableRegistry`.

This design allows for `TypegenSettings` to be constructed independently of any type registry. 
It supports that each derive and attribute can be either recursive or not.

Test added to check the desired behavior.

lexnv commented on .gitignore:
```diff
@@ -7,3 +7,5 @@
 
 Cargo.lock
+
+/artifacts/*.rs
```
nit: Extra endline

lexnv commented on description/src/description.rs:
```diff
@@ -8,12 +8,19 @@ use crate::transformer::Transformer;
 use super::formatting::format_type_description;
 
-pub fn type_description(type_id: u32, type_registry: &PortableRegistry) -> anyhow::Result<String> {
+/// if format is enabled, `format_type_description` is applied to the end result.
```
If this is exposed, I would suggest having more explicit docs for it. Something along the lines of 'Describes the given type ID..' or similar

lexnv commented on description/src/description.rs:
```diff
@@ -8,12 +8,19 @@ use crate::transformer::Transformer;
         ty: &Type<PortableForm>,
         _transformer: &Transformer<String>,
     ) -> anyhow::Result<String> {
+        dbg!(_type_id);
+        dbg!(ty);
```
These could be removed

lexnv commented on description/src/lib.rs:
```diff
@@ -46,61 +82,80 @@ mod tests {
             }"}
         );
     }
-    // todo!("This test with the generics does not fly yet.")
```
Is this a regression introduced by this feature? I would enable it; and then fix it later rather than disabling, what do you think?

lexnv commented on description/src/lib.rs:
```diff
@@ -46,61 +82,80 @@ mod tests {
+        }
+
+        let (type_id, type_registry) = make_type::<Shape<u8>>();
+        dbg!(&type_registry);
```
Do we need this here?

lexnv commented on description/src/transformer.rs:
```diff
@@ -12,7 +12,13 @@ pub struct Transformer<'a, R, S = ()> {
+    /// The `recurse_policy` defines, how to handle cases,
+    /// where a type has been visited before, and is visited again BEFORE a representation of this type could be computed.
+    /// If it returns Some(..) we avoid infinite recursion by returning a concrete value.
+    /// If it returns None, nothing is done in the case of detected recursion and the type falls through.
```
What does `falls through` mean here? Is the recursion continued as the type is "unresolved"?

For which containers could this happen? The blow "this is dangerous" seems a bit scary, might be worth a `# Warning` section

tadeohepperle replied:
Ah yes, the name is not so good, perhaps `recursion_should_continue` is better. On these types we should just let the recursion continue instead of intersecting. That is the idea behind it. I think having this on the containers `Array`, `Sequence`, `Tuple` and `Compact` should be fine, because they cannot recurse infinitely without going through a named type (e.g. a `Variant` or `Composite`) that will be intersected when recursed on.

lexnv commented on description/src/transformer.rs:
```diff
@@ -56,7 +58,10 @@ where
         match self.cache.borrow().get(&type_id) {
             Some(Cached::Recursive) => {
-                return (self.recurse_policy)(type_id, ty, self);
+                dbg!(self.cache.borrow());
```
Same, is this needed?

lexnv commented on description/src/transformer.rs:
```diff
@@ -70,3 +75,46 @@ where
+///
+/// struct B {
+///     id: u8,
+///     others: Vec<B>
```
This is not entirely a valid syntax, does the same problem happen for this? 

```rust
struct B {
  id: u8,
  other: Vec< Box<B> >,
}
```

tadeohepperle replied:
@lexnv What do you mean by "This is not entirely a valid syntax"?

lexnv replied:
Rust would reject recursive types because their `size_of` in bytes wouldn't be known at compile time. And the solution would be to box them, since a pointer always has a known size, depending on the architecture.

I meant that the struct should probably have a boxed::Box(B) instead of recursive type B 😄

lexnv replied:
Ahh I see, its my bad. I was reading it in my mind `struct B { other: B }`. Vec<B> is fine indeed

lexnv commented on description/src/transformer.rs:
````diff
@@ -70,3 +75,46 @@ where
+///     }>
+/// }
+/// ```
+/// But the recursive resolving would get stuck in the middle, reporting recursion.
````
Would something like this solve the issue?

- keep a cache of visited types
- when we encounter a recursion in describing a type, we simply resolve the type to the type's name, iff they match?

```rust
struct B { -> We are describing type B
id: u8,
other: Vec<B> -> Encountered type B name, resolve to the name, because this B type name == B the type name we are currently describing
}
```

jsdw commented on description/src/description.rs:
```diff
@@ -37,19 +49,15 @@ fn ty_description(
+    let prefix = if let TypeDef::Variant(_) = ty.type_def {
+        "enum "
+    } else {
+        ""
```
One thought I had from the prev PR; I think that if we prefix enums with `enum`, that we should also prefix structs with `struct` :)

jsdw commented on description/src/lib.rs:
```diff
@@ -46,61 +87,79 @@ mod tests {
+                Inivisible,
+                Circle(u64),
+                Rect(Compact<u64>, Compact<u64>),
+                Polygon  {
```
Nit; why two spaces between variant name and `{` here and below?

tadeohepperle replied:
That was a bug - fixed it.

jsdw commented on description/src/lib.rs:
```diff
@@ -122,7 +181,7 @@ mod tests {
 
         assert_eq!(
-            type_description(type_id, &type_registry).unwrap(),
+            type_description(type_id, &type_registry, true).unwrap(),
```
Seeing the `Human { .. } ` bit below, and seeing variants like `Foo { .. }` above, gives me another nudge towards a "struct" prefix for clarity, so here we'd have `struct Human {  .. }`. Whatcha think?

tadeohepperle replied:
Hmm, whatever you think fits best. I had explicit `struct` and `enum` prefixes in the web app before and found it littered the entire code a bit, that is why I removed them. But it would be more consistent with Rust syntax.

tadeohepperle replied:
I changed it such that structs now have a `struct` prefix to make it more consistent.

jsdw commented on description/src/transformer.rs:
```diff
@@ -70,3 +72,49 @@ where
+///
+/// # Warning
+///
+/// The safety of the following logic relies on the assumption that ultimately everything resolves down to a primitive or a struct/enum that is in the cache.
+/// It basically just returns true for generic wrapper types.
```
I'm struggling to really understand this `recursion_should_continue` thing; perhaps because I don't grasp the problem well enough!

My understanding so far from reading the above is:

We have type info for types like

```rust
struct A { bees: Vec<B> } 
struct B { id: u8, others: Vec<B> }
```

We'd like to print a description like:

```rust
struct A { 
    bees: Vec<struct B {
        id: u8,
        others: Vec<B>
    }
> 
} 
```

So the first time `Vec<B>` is encountered, we expand B. The second time `Vec<B>` is encountered, we'll be recursing here, and want to instead print `Vec<B>` rather than expanding it.

Naively, to me it feels like this could be achieved by running some function on recursion that takes in the type info, here the info about B, and wants back some string to use. Since we know we are recursing in order to run this function, the function looks at the type name, B, and returns that string here.

> That is because, if the policy you described was applied here, the type which is seen twice in recursion is Vec<B>

You said this above in response to I think alex suggesting the same. But wouldn't the type that's seen twice in recursion be `B` and not `Vec<B>`? (ie, we would recurse into a Vec again, and then see that the type in the vec was B, and then say oohh wait, I've already encountered B before. We wouldn't go into a vec and instantly decide that the vec itself was resurding would we?)

jsdw replied:
Thanks for explaining this!

The bit I missed was that `Vec<B>` is itself a distinct type, and when we hit the second one we'd trigger the "aah we are recursing" thing, and so need to explicitly opt-in to recursing into this sort of type to print it properly without some special case to print `Vec<B>` or whatever.

jsdw commented on description/src/type_example/rust_value.rs:
```diff
@@ -179,13 +190,13 @@ fn ty_example(
             // return a Vec with 2 elements:
             let inner_ty = transformer.resolve_type(def.type_param.id)?;
-            let item_code = ty_example(def.type_param.id, inner_ty, transformer)?; // todo!("Might need CompactMode::Expl")
+            let item_code = ty_example(def.type_param.id, inner_ty, transformer)?;
```
Have you tested some compact encoded types yet? the TODOs are gone but nothing has changed, so I'm wondering how/if they are addressed :)

tadeohepperle replied:
Yes, the different handling of fields that are compact types and fields that just have `#[codec(compact)]`, is addresses here by investigating the `type_name`:
https://github.com/paritytech/scale-typegen/blob/recursive-derives/description/src/type_example/rust_value.rs#L244

I also added a test to confirm this works as expected: 
https://github.com/paritytech/scale-typegen/blob/recursive-derives/description/src/lib.rs#L227

jsdw commented on typegen/src/lib.rs:
```diff
@@ -12,7 +12,21 @@
+#![deny(unused_crate_dependencies)]
+#![deny(missing_docs)]
+
+// The #![deny(unused_crate_dependencies)] requires us to do these for the example to work:
+#[cfg(test)]
+use frame_metadata as _;
+#[cfg(test)]
+use prettyplease as _;
+#[cfg(test)]
+use scale_bits as _;
```
@lexnv would `cargo machete` work better than that unused dependencies check (I saw it mentioned in Subxt but haven't dug into it at all yet :))

jsdw commented on typegen/src/typegen/settings/derives.rs:
```diff
@@ -136,3 +136,161 @@ impl ToTokens for Derives {
+    }
+}
+
+impl DerivesRegistry {
```
Nit; merge this with the above impl block so all impls in one place below the `DerivesRegistry` definition?

jsdw reviewed (approved):
I still think that we should look at the `recursion_should_continue` function to see whether it will always work as intended or whether there's a more general way to handle this, but looks great overall and I don't want to block this from merging any longer!

(I also wonder about whether our documentation not being "inline struct definitions" and instead printing the separate types (ie more natural rust syntax) might be better / more clear (since it's just rust syntax), but would need to try both and see what felt best, and in case case that'd be a separate PR I expect!)

tadeohepperle commented:
I got rid of the `recursion_should_continue` by delegating this responsibility to the `recurse_policy` or a transformer. This now returns an Option. When None is returned, we just continue recursing. Now we look at the type and if it does not have an identifier, we keep recursing when building a type description. This covers the same cases, the  `recursion_should_continue` function covered before. 

I also added better support for printing nested generic types and added tests for this.
